runtime/src/main/java/org/corfudb/protocols/wireprotocol/ILogData.java

```diff
@@ -67,13 +66,28 @@ public void close() {
         }
     }
 
-    void releaseBuffer();
+    /**
+     * Get the serialization handle of this entry that manages
+     * the lifetime of the serialized copy.
+     *
+     * @param metadata whether metadata needs to be serialized
+     * @return a serialization handle of this entry
+     */
+    default SerializationHandle getSerializedForm(boolean metadata) {
+        return new SerializationHandle(this, metadata);
+    }
 
-    void acquireBuffer();
+    /**
+     * Release the serialization buffer.
+     */
+    void releaseBuffer();
 
-    default SerializationHandle getSerializedForm() {
-        return new SerializationHandle(this);
-    }
+    /**
+     * Acquire the serialization buffer.
+     *
+     * @param metadata whether metadata needs to be serialized
+     */
+    void acquireBuffer(boolean metadata);
```

Comment:
I think the metadata serialization is cheap enough to actually do it twice. 
I would just serialize both metadata + payload here. Then after getting a token just call `doSerializeMetadataInternal` with the last ByteBuf of the Composite buffer. It will overwrite it with the new token information. 

Instead of adding the boolean here, you can just introduce a method `updateMetadata` that will rewrite the metadata portion.

Reply:
The reason I added this flag is that sometimes, metadata is even larger compared to the payload (e.g. logunit cache oom issue we used to fix), and also when the token is not being set, the metadata is in-complete, serializing the metadata makes the meaning of the serialized cache a bit unclear. What do you think?

Reply:
I see.